File: python/tests/test_self_ip_cache.py
"""Memoised self inner products, the oneSidedDenom <X,X> skip, and the
all-r = 1 direct inner-product route.

The contract under test: a broadcast (one context density against many
query densities) returns exactly the values that fresh scalar calls
return, under both normalisations and on both the Bulger and Möbius
routes; the r = 1 direct route computes the identical quantity to the
generic MA path; and the memoised values never leak across settings
that change them.
"""
import numpy as np
import pytest

import mpt
from mpt._defaults import resolve_truncation_sigmas
from mpt._tensor import cosine as C


RNG = np.random.default_rng(7)


def _r1_density(pitches, times, sigma=(0.25, 0.08), is_per=(False, False),
                period=(0.0, 0.0)):
    return mpt.build_exp_tens(
        [np.asarray(pitches, dtype=float).reshape(1, -1),
         np.asarray(times, dtype=float).reshape(1, -1)],
        None, list(sigma), [1, 1], [False, False], list(is_per),
        list(period), verbose=False,
    )


# ------------------------------------------------------------------
#  Broadcast == fresh scalar calls (cache transparency), Bulger route
# ------------------------------------------------------------------

@pytest.mark.parametrize("norm", ["cosine", "oneSidedDenom"])
def test_broadcast_matches_fresh_scalars_r1(norm):
    N = 150
    P = RNG.uniform(40, 90, N)
    T = np.sort(RNG.uniform(0, 40, N))
    qP = np.array([60.0, 63.0, 68.0])
    qT = np.array([0.0, 1.0, 3.0])

    dX = _r1_density(P, T)
    dYs = [_r1_density(qP + k * 0.7, qT + k * 1.3) for k in range(6)]
    batch = mpt.cos_sim_exp_tens(dX, dYs, normalize=norm, verbose=False)

    fresh = np.array([
        mpt.cos_sim_exp_tens(
            _r1_density(P, T), _r1_density(qP + k * 0.7, qT + k * 1.3),
            normalize=norm, verbose=False)
        for k in range(6)
    ])
    np.testing.assert_allclose(batch, fresh, rtol=0, atol=1e-12)


# ------------------------------------------------------------------
#  Broadcast == fresh scalar calls on the Möbius (orbit) route
# ------------------------------------------------------------------

def _r2_density(vals, sigma=1.0, n_events=None):
    vals = np.asarray(vals, dtype=float)
    return mpt.build_exp_tens(
        [vals], None, [sigma], [2], [True], [False], [0.0],
        verbose=False,
    )


@pytest.mark.parametrize("norm", ["cosine", "oneSidedDenom"])
def test_broadcast_matches_fresh_scalars_mobius(norm):
    # r = 2, relative, K large enough that the dispatcher's Möbius
    # route is exercised when forced; forcing keeps the test route-
    # deterministic while the equality contract stays the point.
    X = RNG.uniform(0, 24, (2, 30))
    Ys = [RNG.uniform(0, 24, (2, 6)) for _ in range(4)]

    def build(v):
        return mpt.build_exp_tens([v], None, [1.0], [2], [True], [False],
                                  [0.0], verbose=False)

    dX = build(X)
    dYs = [build(Y) for Y in Ys]
    batch = np.array([
        mpt.cos_sim_exp_tens(dX, dY, method="mobius", normalize=norm,
                             verbose=False)
        for dY in dYs
    ])
    fresh = np.array([
        mpt.cos_sim_exp_tens(build(X), build(Y), method="mobius",
                             normalize=norm, verbose=False)
        for Y in Ys
    ])
    np.testing.assert_allclose(batch, fresh, rtol=0, atol=1e-12)


# ------------------------------------------------------------------
#  r = 1 direct route == generic MA path, all abs mode combinations
# ------------------------------------------------------------------

@pytest.mark.parametrize("is_per,period", [
    ((False, False), (0.0, 0.0)),
    ((True, False), (12.0, 0.0)),
    ((True, True), (12.0, 4.0)),
])
def test_r1_direct_matches_generic(is_per, period):
    d1 = _r1_density(RNG.uniform(0, 24, 31), RNG.uniform(0, 8, 31),
                     sigma=(0.4, 0.1), is_per=is_per, period=period)
    d2 = _r1_density(RNG.uniform(0, 24, 17), RNG.uniform(0, 8, 17),
                     sigma=(0.4, 0.1), is_per=is_per, period=period)
    ts = resolve_truncation_sigmas(None)
    fast = C._ip_r1_direct(
        d1.u_perm, d1.w_j, d1.n_j, d2.v_comb, d2.wv_comb, d2.n_k,
        2, d1.sigma, d1.is_rel, d1.is_per, d1.period,
        truncation_sigmas=ts, wrap=list(d1.wrap),
    )
    slow = C._ip_full_ma(
        d1.u_perm, d1.w_j, d1.n_j, d2.v_comb, d2.wv_comb, d2.n_k,
        2, d1.r, d1.sigma, d1.is_rel, d1.is_per, d1.period,
        truncation_sigmas=ts, inner_r=None, wrap=list(d1.wrap),
    )
    assert fast == pytest.approx(slow, rel=1e-14, abs=0.0)


# ------------------------------------------------------------------
#  oneSidedDenom skips <X,X>; cosine refuses a skipped <X,X>
# ------------------------------------------------------------------

def test_pairwise_need_xx_false_returns_none_xx():
    dX = _r1_density(RNG.uniform(40, 90, 50),
                     np.sort(RNG.uniform(0, 20, 50))).pruned()
    dY = _r1_density([60.0, 64.0, 67.0], [0.0, 1.0, 2.0]).pruned()
    ip_xy, ip_xx, ip_yy = C._cos_sim_exp_tens_ma_pairwise(
        dX, dY, verbose=False, need_xx=False)
    assert ip_xx is None
    assert np.isfinite(ip_xy) and np.isfinite(ip_yy)
    # The value oneSidedDenom produces from the reduced triple matches
    # the full-triple computation.
    full = C._cos_sim_exp_tens_ma_pairwise(dX, dY, verbose=False)
    assert ip_xy == full[0] and ip_yy == full[2]
    with pytest.raises(ValueError):
        C._finalise_normalisation(ip_xy, None, ip_yy, "cosine")


def test_cached_self_ip_is_reused_and_keyed():
    dX = _r1_density(RNG.uniform(40, 90, 40),
                     np.sort(RNG.uniform(0, 20, 40))).pruned()
    dY = _r1_density([60.0, 64.0], [0.0, 1.0]).pruned()
    C._cos_sim_exp_tens_ma_pairwise(dX, dY, verbose=False)
    key = C._self_ip_cache_key("bulger", None, None)
    assert key in dX._self_ip_cache and key in dY._self_ip_cache
    cached = dX._self_ip_cache[key]
    # Second call reuses the cached value (identity of the float, and
    # the triple slot equals it exactly).
    trip = C._cos_sim_exp_tens_ma_pairwise(dX, dY, verbose=False)
    assert trip[1] == cached
    # A different truncation budget is a different key: no leak.
    C._cos_sim_exp_tens_ma_pairwise(dX, dY, verbose=False,
                                    truncation_sigmas=6.0)
    key6 = C._self_ip_cache_key("bulger", 6.0, None)
    assert key6 in dX._self_ip_cache
    assert key6 != key


def test_pruned_is_memoised():
    d = _r1_density(RNG.uniform(40, 90, 10), np.sort(RNG.uniform(0, 5, 10)))
    assert d.pruned() is d.pruned()

File: matlab/tests/test_self_ip_cache.m
%% test_self_ip_cache.m — memoised self IPs, oneSidedDenom skip, r = 1 route
%
%  Tests for the self-inner-product memoisation, the 'oneSidedDenom'
%  <X,X> skip, and the direct all-r = 1 inner-product route in
%  cosSimExpTens. Covers:
%    - Sweep (scalar-vs-cell broadcast) returns exactly the values that
%      fresh scalar calls return, under both normalisations (the memo
%      is a pure optimisation; values are unchanged).
%    - The same equality on the Möbius route (method = 'mobius'
%      forced), exercising the orbit-side memo and choices key.
%    - The cache-carrying outputs: a manual scalar loop threading
%      [s, densRef] = cosSimExpTens(densRef, ...) matches the sweep,
%      and the returned struct carries a populated 'selfIP' field.
%    - Cross-language goldens: r = 1 two-attribute similarities on
%      formula-based inputs match the Python toolbox to 1e-12 relative,
%      in all three abs-mode combinations and both normalisations
%      (this pins the r = 1 direct route against the independently
%      validated Python path), plus one Möbius abs r = 3 golden.
%    - The cache-carrying outputs are refused outside the
%      density-struct scalar form.
%
%  Standalone-runnable. When invoked from test_mpt.m the existing
%  `results` cell is appended to; when run alone, results print on the
%  fly and a summary follows.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end

%% ---- Shared deterministic inputs (formula-based; identical in Python) ----

iVec = 1:12;
P = 50 + mod(7 * iVec.^2, 37);
T = cumsum(mod(3 * iVec, 5) / 4 + 0.2);
qP = [60 63.5 68.25];
qT = [0.5 1.25 2.0];

modesPer  = {[false false], [true false], [true true]};
modesPrd  = {[0 0], [12 0], [12 4]};
modesName = {'nonper', 'perA', 'perAB'};

% Python-computed goldens (goldens_inf.py; formula-based inputs above),
% computed with truncation_sigmas = inf --- the un-truncated
% accuracy-floor path that mptTestIsolateDefaults pins as the suite
% baseline (the factory default of 6 is a ~6-sig-fig approximation,
% which a 1e-12 golden would fail by construction; the perAB mode
% coincides at either setting because every wrapped distance there
% falls inside the effective truncation window).
golden = struct();
golden.nonper.cosine        = [6.682242385236272e-03, 1.968354486993972e-03, 1.730893616999029e-05];
golden.nonper.oneSidedDenom = [1.364172585654229e-02, 4.018374484498330e-03, 3.533600676040842e-05];
golden.perA.cosine          = [5.737534202374079e-02, 1.245711904636613e-01, 2.814032744888095e-01];
golden.perA.oneSidedDenom   = [1.306592271612683e-01, 2.836824130095172e-01, 6.408316372239535e-01];
golden.perAB.cosine         = [3.162529342551098e-01, 2.510893754426644e-01, 4.565179404859586e-01];
golden.perAB.oneSidedDenom  = [8.376450969950919e-01, 6.650492737482304e-01, 1.209158787535208e+00];

%% ---- Sweep == fresh scalars, plus cross-language goldens (r = 1) ----

for mi = 1:3
    isPerM = modesPer{mi};
    prdM   = modesPrd{mi};
    dX = buildExpTens({P, T}, [], [0.9 0.35], [1 1], [false false], ...
        isPerM, prdM, 'verbose', false);
    dYs = cell(1, 3);
    for k = 0:2
        dYs{k+1} = buildExpTens({qP + 0.6*k, qT + 0.9*k}, [], ...
            [0.9 0.35], [1 1], [false false], isPerM, prdM, ...
            'verbose', false);
    end
    for nrmC = {'cosine', 'oneSidedDenom'}
        nrm = nrmC{1};
        sweep = cell2mat(cosSimExpTens(dX, dYs, ...
            'normalize', nrm, 'verbose', false));
        fresh = zeros(1, 3);
        for k = 1:3
            dXf = buildExpTens({P, T}, [], [0.9 0.35], [1 1], ...
                [false false], isPerM, prdM, 'verbose', false);
            fresh(k) = cosSimExpTens(dXf, dYs{k}, ...
                'normalize', nrm, 'verbose', false);
        end
        results{end+1,1} = sprintf( ...
            'selfIP.r1 %s %s: sweep matches fresh scalars (0 diff)', ...
            modesName{mi}, nrm); %#ok<*SAGROW>
        results{end,2}   = max(abs(sweep - fresh)) == 0;

        g = golden.(modesName{mi}).(nrm);
        results{end+1,1} = sprintf( ...
            'selfIP.r1 %s %s: matches Python golden (1e-12 rel)', ...
            modesName{mi}, nrm);
        results{end,2}   = max(abs(sweep - g) ./ max(abs(g), 1e-300)) < 1e-12;
    end
end

%% ---- Cache-carrying outputs: threaded loop matches sweep ----

dX = buildExpTens({P, T}, [], [0.9 0.35], [1 1], [false false], ...
    [false false], [0 0], 'verbose', false);
dYs = cell(1, 3);
for k = 0:2
    dYs{k+1} = buildExpTens({qP + 0.6*k, qT + 0.9*k}, [], [0.9 0.35], ...
        [1 1], [false false], [false false], [0 0], 'verbose', false);
end
sweep = cell2mat(cosSimExpTens(dX, dYs, 'verbose', false));
dRef = dX;
sLoop = zeros(1, 3);
for k = 1:3
    [sLoop(k), dRef] = cosSimExpTens(dRef, dYs{k}, 'verbose', false);
end
results{end+1,1} = 'selfIP.threaded loop via 2nd output matches sweep (0 diff)';
results{end,2}   = max(abs(sLoop - sweep)) == 0;
results{end+1,1} = 'selfIP.returned struct carries a populated selfIP field';
results{end,2}   = isfield(dRef, 'selfIP') && isstruct(dRef.selfIP) ...
    && numel(dRef.selfIP.keys) >= 1;

%% ---- Möbius route: sweep == fresh, and abs r = 3 golden ----

jVec = 1:15;
xFlat = 100 * mod(11 * jVec.^2, 19) / 19 + 10 * jVec;
yFlat = 100 * mod(13 * jVec.^2, 23) / 23 + 9 * jVec;
% MATLAB reshape is column-major; NumPy's is row-major. Transpose the
% 3-by-5 column-major reshape to obtain NumPy's (5, 3) row-major array.
X3 = reshape(xFlat, 3, 5).';
Y3 = reshape(yFlat, 3, 5).';
bld3 = @(v) buildExpTens({v}, [], 8.0, 3, false, false, 0, 'verbose', false);

goldMob = struct('cosine', 1.566062218777721e-01, ...
                 'oneSidedDenom', 1.497935199140709e-01);
for nrmC = {'cosine', 'oneSidedDenom'}
    nrm = nrmC{1};
    dX3 = bld3(X3);
    sMemo1 = cosSimExpTens(dX3, bld3(Y3), 'method', 'mobius', ...
        'normalize', nrm, 'verbose', false);
    sFresh = cosSimExpTens(bld3(X3), bld3(Y3), 'method', 'mobius', ...
        'normalize', nrm, 'verbose', false);
    results{end+1,1} = sprintf( ...
        'selfIP.mobius %s: repeated-operand call matches fresh (0 diff)', nrm);
    results{end,2}   = abs(sMemo1 - sFresh) == 0;
    results{end+1,1} = sprintf( ...
        'selfIP.mobius %s: matches Python golden (1e-9 rel)', nrm);
    results{end,2}   = abs(sFresh - goldMob.(nrm)) ...
        / abs(goldMob.(nrm)) < 1e-9;
end

%% ---- Cache-carrying outputs refused outside the scalar struct form ----

okListRefused = false;
try
    [~, ~] = cosSimExpTens(dX, dYs, 'verbose', false); %#ok<ASGLU>
catch ME
    okListRefused = strcmp(ME.identifier, ...
        'cosSimExpTens:selfIpOutputsUnavailable');
end
results{end+1,1} = 'selfIP.2nd output in list mode raises selfIpOutputsUnavailable';
results{end,2}   = okListRefused;

%% ---- Summary (standalone only) ----

if standalone
    nPass = sum(cell2mat(results(:,2)));
    fprintf('\ntest_self_ip_cache: %d/%d passed\n', nPass, size(results, 1));
    for ri = 1:size(results, 1)
        if ~results{ri,2}
            fprintf('  FAIL: %s\n', results{ri,1});
        end
    end
end
